File: src/examples/c_lib.clj
;; IMPORTANT!
;; (System/setProperty "jna.library.path" "your/lib/path/here")
;; Must be done before calling loadlib, preferrably on command line


(ns examples.c-lib
  (:use [clj-native.direct :only [defclib loadlib typeof]]
        [clj-native.structs :only [byref byval]]
        [clj-native.callbacks :only [callback]]))

(defclib
  my-lib-name ;;  c_lib
  (:libname "c_lib")
  (:structs
   (struct1 :x int :y char :k float)
   (struct2 :ll longlong :s1ByValue struct1)
   (circle1 :c2 circle2*)
   (circle2 :c1 circle1*)
   (c-list :data void* :next c-list*) ;; can't really name it list
   (packed :s1 short :s2 short))
  (:unions
   (splitint :theint int :packed packed))
  (:callbacks
   (add-cb [int int] int)
   (reply-callback [void* void* i32] void)
   (void-param-callback [void*] void))
  (:functions
   (add [int int] int)
   (call-add-callback call_add_callback [add-cb int int] int)
   (call-void-param-callback call_void_param_callback [void-param-callback void*] void)
   (get-ptr get_ptr [] void*)
   (call-reply-callback call_reply_callback [reply-callback void* char* i32] void)
   (count-bytes count_bytes [char*] long)
   (addOneToStructByReference [struct1*] struct1*)
   (addOneToStructByValue [struct1] struct1)
   (addOneToStructTwoByValue [struct2] struct2)
   (returnsConstantString [] constchar*)
   (returnsConstantWString [] constwchar_t*)
   (addOneToUnionIntByValue [splitint] splitint)
   (addOneToUnionIntByReference [splitint*])))

  (defn main
    []
    (loadlib my-lib-name)
    ;; (loadlib c_lib)
    (let [cb (callback add-cb (fn [x y]
                                (println "in callback!")
                                (+ x y)))
          rcb (callback reply-callback
                        (fn [ptr buf size]
                          (let [bb (.getByteBuffer buf 0 10000)
                                n-bytes (count-bytes bb)]
                            (println "in reply callback! size is " size n-bytes bb))))
          vcb (callback void-param-callback
                        (fn [vp]
                          (println "The pointer is" vp)))
          vptr (get-ptr)
          s1val (byval struct1)
          s1ref (byref struct1)
          s2val (byval struct2)
          ;; set alignment to match the C code (packed)
          splitintval (byval splitint com.sun.jna.Structure/ALIGN_NONE)
          splitintref (byref splitint com.sun.jna.Structure/ALIGN_NONE)
          countme (java.nio.ByteBuffer/allocate 10000)]
      (dotimes [_ 10000] (.put countme (byte 0)))
      (.rewind countme)
      (dotimes [_ 100] (.put countme (byte 1)))
      (.rewind countme)
      (println "count 100 bytes:" (count-bytes countme))
      (call-void-param-callback vcb vptr)
      (println "vptr is" vptr)
      (println "Result of add(10, 35):" (add 10 35))
      (println "Result of call_add_callback(cb, 10, 78):"
               (call-add-callback cb 10 78))
      (println "Result of call_add_callback(cb, 90, 78):"
               (call-add-callback cb 90 78))
      (println "Passing struct1 by value")
      (println s1val)
      (println (addOneToStructByValue s1val))
      (println s1val)
      (println "Passing struct1 by reference")
      (println s1ref)
      (println (addOneToStructByReference s1ref))
      (println s1ref)
      (println "testing reply callback")
      (println "Result of call_reply_callback():"
               (call-reply-callback rcb vptr countme 1))
      (println "Passing struct2 by value")
      (println s2val)
      (println (addOneToStructTwoByValue s2val))
      (println s2val)
      (println (returnsConstantString))
      (println (returnsConstantWString))
      (.setType splitintval Integer/TYPE)
      (set! (.theint splitintval) 66000)
      (.setType splitintref Integer/TYPE)
      (set! (.theint splitintref) 66000)
      (println "Passing splitint by value")
      (println (.theint splitintval))
      (let [ret (addOneToUnionIntByValue splitintval)
            _ (.readField ret "packed") ;; force read. Should this really be necessary?
            ;; _ (.setType ret (typeof packed :val)) ;; I wish this forced a read
            s1 (.s1 (.packed ret))
            s2 (.s2 (.packed ret))]
        (println (str " s1: " s1
                      ", s2: " s2)))
      (println "Passing splitint by reference")
      (println (.theint splitintref))
      (addOneToUnionIntByReference splitintref)
      (.readField splitintref "packed") ;; force read
      ;; (.setType splitintref (typeof packed :val))
      (println (str " s1: " (.s1 (.packed splitintref))
                    ", s2: " (.s2 (.packed splitintref))))))

(comment
  (main)
  )

File: src/examples/c_lib.c
#include <stdlib.h>
#include <stdio.h>
#include <wchar.h>
#include <string.h>

#ifndef _WIN32
  #define EXPORT
#else
  #define EXPORT __declspec(dllexport)
#endif

/* globals */
EXPORT int globalInt = 10;
EXPORT const char* globalString = "Hello Globals!";

/* structures and unions */
typedef struct struct1_struct {
  int x;
  char y; /* alignment issues? */
  float k;
} struct1;

typedef struct struct2_struct {
  long long ll;
  struct1 s1ByValue;
} struct2;

#ifdef _WIN32
#pragma pack(push, 1)
typedef struct packed_struct {
  short s1;
  short s2;
} packed;
#pragma pack(pop)
#else // assumes gcc
typedef struct packed_struct {
  short s1;
  short s2;
} packed __attribute__ ((aligned (1)));
#endif

typedef union splitint_union {
  int the_int;
  packed p;
} splitint;

typedef union either_union {
  struct1* s1;
  struct2* s2;
} either;

/* structures with circular references */
typedef struct circle2_struct circle2;

typedef struct circle1_struct {
  circle2* c2;
} circle1;

struct circle2_struct {
  circle1* c1;
};

/* structure with reference to own type */
typedef struct list_struct list;

struct list_struct {
  void* data;
  list* next;
};

/* functions and callbacks */
EXPORT int add(int x, int y) {
  return x + y;
}

EXPORT size_t count_bytes(unsigned char* buf) {
  size_t count = 0;

  while(buf[count]) {
    ++count;
  }

  return count;
}

typedef int (*add_callback)(int, int);

typedef void (*void_param_callback)(void* vp);

struct ReplyAddress {int a; int b;};

typedef void (*reply_callback)(struct ReplyAddress *inReplyAddr, char* inBuf, int inSize);

EXPORT int call_add_callback(add_callback cb, int x, int y) {
  return cb(x, y);
}

EXPORT void call_void_param_callback(void_param_callback cb, void* vp) {
  cb(vp);
}

EXPORT void* get_ptr(void)
{
  return malloc((size_t)100);
}

EXPORT void call_reply_callback(reply_callback cb, struct ReplyAddress *inReplyAddr, char* inBuf, int inSize) {
  cb(inReplyAddr,inBuf,inSize);
}

EXPORT struct2 addOneToStructTwoByValue(struct2 s2)
{
  struct2 ret;
  ret.ll = s2.ll + 1;
  ret.s1ByValue.x = s2.s1ByValue.x + 1;
  ret.s1ByValue.y = s2.s1ByValue.y + 1;
  ret.s1ByValue.k = s2.s1ByValue.k + 1;
  return ret;
}

EXPORT struct1 addOneToStructByValue(struct1 s1)
{
  struct1 ret;
  ret.x = s1.x + 1;
  ret.y = s1.y + 1;
  ret.k = s1.k + 1;
  return ret;
}

EXPORT void addOneToStructByReference(struct1* s1)
{
  ++s1->x;
  ++s1->y;
  ++s1->k;
}

EXPORT const char* returnsConstantString()
{
  return "This string should be safe to read as const char*";
}

EXPORT const wchar_t* returnsConstantWString()
{
  return L"This string should be safe to read as const wchar_t*";
}

EXPORT splitint addOneToUnionIntByValue(splitint s1)
{
  splitint ret;
  ret.the_int = s1.the_int + 1;
  return ret;
}

EXPORT void addOneToUnionIntByReference(splitint* s1)
{
  ++s1->the_int;
}

